File: server/src/Connection/Messages/PlayerInfo.cs
using System.Text.Json.Serialization;

namespace GameServer.Connection;

public record PlayerInfo : Message
{
    [JsonPropertyName("messageType")]
    public override string MessageType { get; init; } = "PlayerInfo";

    [JsonPropertyName("token")]
    public string Token { get; init; } = "";

    public record weapon
    {
        [JsonPropertyName("attackSpeed")]
        public double AttackSpeed { get; init; }

        [JsonPropertyName("bulletSpeed")]
        public double BulletSpeed { get; init; }

        [JsonPropertyName("isLaser")]
        public bool IsLaser { get; init; }

        [JsonPropertyName("antiArmor")]
        public bool AntiArmor { get; init; }

        [JsonPropertyName("damage")]
        public int Damage { get; init; } = 0;

        [JsonPropertyName("maxBullets")]
        public int MaxBullets { get; init; } = 0;

        [JsonPropertyName("currentBullets")]
        public int CurrentBullets { get; init; } = 0;
    }

    [JsonPropertyName("weapon")]
    public weapon Weapon { get; init; } = new();

    public record armor 
    {
        [JsonPropertyName("canReflect")]
        public bool CanReflect { get; init; }

        [JsonPropertyName("armorValue")]
        public int ArmorValue { get; init; } = 0;

        [JsonPropertyName("health")]
        public int Health { get; init; } = 0;

        [JsonPropertyName("gravityField")]
        public bool GravityField { get; init; }

        [JsonPropertyName("knife")]
        public string Knife { get; init; } = "";

        [JsonPropertyName("dodgeRate")]
        public double DodgeRate { get; init; }
    }

    [JsonPropertyName("armor")]
    public armor Armor { get; init; } = new();

    [JsonPropertyName("skills")]
    public List<skill> Skills { get; init; } = new();

    public record skill
    {
        [JsonPropertyName("name")]
        public string Name { get; init; } = "";

        [JsonPropertyName("maxCooldown")]
        public int MaxCooldown { get; init; } = 0;

        [JsonPropertyName("currentCooldown")]
        public int CurrentCooldown { get; init; } = 0;

        [JsonPropertyName("isActive")]
        public bool IsActive { get; init; }
    }

    [JsonPropertyName("position")]
    public position Position { get; init; } = new();

    public record position
    {
        [JsonPropertyName("x")]
        public double X { get; init; }

        [JsonPropertyName("y")]
        public double Y { get; init; }

        [JsonPropertyName("angle")]
        public double Angle { get; init; }
    }

}

File: sdk/python/src/agent/player_info.py
from typing import List, Literal

from .position import Position


class WeaponInfo:
    def __init__(
        self,
        attack_speed: float,
        bullet_speed: float,
        is_laser: bool,
        anti_armor: bool,
        damage: int,
        max_bullets: int,
        current_bullets: int,
    ):
        self.attack_speed = attack_speed
        self.bullet_speed = bullet_speed
        self.is_laser = is_laser
        self.anti_armor = anti_armor
        self.damage = damage
        self.max_bullets = max_bullets
        self.current_bullets = current_bullets

    def __str__(self):
        return (
            f"Weapon(attack_speed={self.attack_speed}, bullet_speed={self.bullet_speed}, "
            f"is_laser={self.is_laser}, anti_armor={self.anti_armor}, damage={self.damage}, "
            f"max_bullets={self.max_bullets}, current_bullets={self.current_bullets})"
        )


class ArmorInfo:
    def __init__(
        self,
        can_reflect: bool,
        armor_value: int,
        health: int,
        gravity_field: bool,
        knife: Literal["NOT_OWNED", "AVAILABLE", "ACTIVE", "BROKEN"],
        dodge_rate: float,
    ):
        self.can_reflect = can_reflect
        self.armor_value = armor_value
        self.health = health
        self.gravity_field = gravity_field
        self.knife = knife
        self.dodge_rate = dodge_rate

    def __str__(self):
        return (
            f"Armor(can_reflect={self.can_reflect}, armor_value={self.armor_value}, "
            f"health={self.health}, gravity_field={self.gravity_field}, knife={self.knife}, "
            f"dodge_rate={self.dodge_rate})"
        )


class SkillInfo:
    def __init__(
        self, name: str, max_cooldown: int, current_cooldown: int, is_active: bool
    ):
        self.name = name
        self.max_cooldown = max_cooldown
        self.current_cooldown = current_cooldown
        self.is_active = is_active

    def __str__(self):
        return (
            f"Skill(name={self.name}, max_cooldown={self.max_cooldown}, "
            f"current_cooldown={self.current_cooldown}, is_active={self.is_active})"
        )


class PlayerInfo:
    def __init__(
        self,
        token: str,
        weapon: WeaponInfo,
        armor: ArmorInfo,
        skills: List[SkillInfo],
        position: Position,
    ):
        self.token = token
        self.weapon = weapon
        self.armor = armor
        self.skills = skills
        self.position = position

    def __str__(self) -> str:
        return f"PlayerInfo{{token: {self.token}, weapon: {self.weapon}, armor: {self.armor}, skill: {self.skills}, position: {self.position}}}"
